### PR TITLE
include support for hooks

### PR DESCRIPTION
**Pull request by inode64**
Include help too

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,6 +62,18 @@ comparison function. For anything else, see [Package configuration](#package-con
 For packages that will not work with currently heuristic checking, a configuration file named
 `livecheck.json` can be placed in the directory alongside the ebuild.
 
+## hook parameters
+
+1. name: Name of the hook script to execute.
+2. action: Action type, e.g., `post` or `pre`
```
Comment (Tatsh) [inline README.md:68]:
```suggestion
- Action type, e.g. `post` or `pre`
```

Comment (Tatsh) [inline README.md:68]:
```suggestion
- Action type, e.g. `post` or `pre`.
```

File: README.md [hunk 2]
```diff
@@ -60,6 +62,18 @@ comparison function. For anything else, see [Package configuration](#package-con
...
+3. search_dir: Root portage directory, e.g., `/usr/portage`
+4. cp: Category and ebuild name, e.g., `dev-lang/php`
+5. old_version: Previous version, e.g., `8.2.32-r2`
+6. new_version: New version, e.g., `8.2.33`
+7. old_sha: (Optional) SHA hash of the old version.
+8. new_sha: (Optional) SHA hash of the new version.
+9. hash_date: (Optional) Date associated with the hash.
```
Comment (Tatsh) [inline README.md:75]:
```suggestion
- Category and package name, e.g. `dev-lang/php`.
- Previous version, e.g. `8.2.32-r2`.
- New version, e.g. `8.2.33`.
- SHA hash of the old version. Optional.
- SHA hash of the new version. Optional.
- Date associated with the hash. Optional.
```

File: livecheck/main.py [hunk 3]
```diff
@@ -334,7 +334,7 @@ def _replace_format(match: Match[str], replacements: dict[str, str]) -> str:
 
 def do_main(*, auto_update: bool, keep_old: bool, cat: str, ebuild_version: str, pkg: str,
             search_dir: str, settings: LivecheckSettings, top_hash: str, hash_date: str, url: str,
-            git: bool) -> None:
+            git: bool, hook: click.File | None) -> None:
```
Comment (Tatsh) [inline livecheck/main.py:337]:
```suggestion
            git: bool, hook: str | None) -> None:
```

File: livecheck/main.py [hunk 4]
```diff
@@ -408,6 +408,10 @@ def do_main(*, auto_update: bool, keep_old: bool, cat: str, ebuild_version: str,
                     sp.run(('mv', ebuild, new_filename), check=True)
             with open(new_filename, 'w') as f:
                 f.write(content)
+            if hook and sp.run((hook.name, 'pre', search_dir, cp, str_old_version, str_new_version,
+                                old_sha, new_sha, hash_date),
+                               check=False).returncode != 0:
+                logger.error(f'Error running hook {hook.name}')
```
Comment (Tatsh) [inline livecheck/main.py:414]: Shouldn't this exit here?

```suggestion
                click.echo(f'Error running hook {hook.name}.', err=True)
                raise click.Abort
```

File: livecheck/main.py [hunk 5]
```diff
@@ -456,6 +460,10 @@ def do_main(*, auto_update: bool, keep_old: bool, cat: str, ebuild_version: str,
                     sp.run(('pkgdev', 'commit'), cwd=os.path.join(search_dir, cp), check=True)
                 except sp.CalledProcessError:
                     logger.error(f'Error committing {new_filename}')
+            if hook and sp.run((hook.name, 'post', search_dir, cp, str_old_version, str_new_version,
+                                old_sha, new_sha, hash_date),
+                               check=False).returncode != 0:
+                logger.error(f'Error running hook {hook.name}')
```
Comment (Tatsh) [inline livecheck/main.py:466]:
```suggestion
                click.echo(f'Error running hook {hook.name}.', err=True)
                raise click.Abort
```

**Comment by Tatsh** [comment]
I think it would be better to read hooks in order from a directory like other things do. You can use [platformdirs](https://platformdirs.readthedocs.io/en/latest/) and read from like: `sorted(listdir(platformdirs.user_config_path('xirvik', ensure_exists=True) / 'pre'))` (or `'post'`). Then for every single file (non-recursive), if the file is executable, run it.

This change would allow any amount of hooks to exist.

**Review by Tatsh — CHANGES_REQUESTED** [review]

**Comment by inode64** [comment]
I have included a directory of hooks so that it can be managed from github and included in the repository. Because if it is in the configuration it could not be included clearly in the overlay.
See: 
https://github.com/inode64/inode64-overlay/tree/main/scripts/hooks

**Pull request merged**